### PR TITLE
Examples: Add GPU Progressive Lightmapping Example

### LINKED ISSUE #14048: Soft Shadows via Light Source Sampling
The idea here is that we use the fastest shadow algorithm possible, and then accumulate it across frames.  To get accurate accumulated soft shadows, we should jitter the light origin within the volume of the light.  For a lights of any type we jitter in their volume in a Poisson-like fashion.  This will create physically correct penumbra if you accumulate over time.

We can either accumulate in the main frame buffer or we could accumulate in a shadow buffer.

Sketchfab is currently jittering their light sources in order to get soft shadows when the scene is static.  We have also recently implemented a basic light jitter approach in Clara.io and it gives good result.  We should take it to the next level and integrate it with core Three.JS.

### PR DESCRIPTION
**Pull request by zalo**
Related issue: Fixed #14048

**Description**

This PR adds a light-weight example for generating silky smooth, physically-correct lightmaps over time by jittering directional lights + shadows (1st bounce only).

The lightmapping accumulation step for this scene takes between 0.2ms and 0.4ms per frame on my machine (GTX 1080).

While the setup currently generates physically accurate soft shadows and ambient occlusion, the technique could generalize to multibounce global illumination/radiosity with the right kind of depth-peeling "light sources".

Check out the live deployment here:
https://zalo.github.io/three.js/examples/?q=shadowmap#webgl_shadowmap_progressive

image

The [potpack module](https://mapbox.github.io/potpack/) is a (very) small utility for packing multiple mesh's UVs into a unified texture atlas for the lightmap.

### CODE REVIEW

**Comment by Mugen87** [comment]
> examples/models/obj/ShadowmappableMesh.obj 

Do you mind using a `glb` file instead?

**Comment by Mugen87** [comment]
The E2E test for `webgl_shadowmap_progressive` is still failing. Maybe you have to regenerate the screenshot.

For testing: https://raw.githack.com/zalo/three.js/progressive-lightmapping/examples/webgl_shadowmap_progressive.html

File: examples/jsm/misc/ProgressiveLightMap.js [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+ * lighting samples.
+ *
+ * This should begin accumulating lightmaps which apply to
+ * your objects, so you can start jittering lighting to achieve
+ * the texture-space effect you're looking for.
+ *
+ * @param {WebGLRenderer} renderer A WebGL Rendering Context
+ * @param {number} res The side-long dimension of you total lightmap
+ */
+var ProgressiveLightMap = function ( renderer, res = 1024 ) {
```
Comment (Mugen87) [inline examples/jsm/misc/ProgressiveLightMap.js:20]: Do you mind writing this code as a ES6 class?

Comment (zalo) [reply examples/jsm/misc/ProgressiveLightMap.js]: Interesting; I'm still learning when it comes to Modern Javascript Syntax, so hopefully I did that right...

Comment (Mugen87) [reply examples/jsm/misc/ProgressiveLightMap.js]: It looks good! 👍

**Comment by zalo** [comment]
Happy to add additional stylization, tests, API changes, or aspects to the scene (if desired).  Don’t hesitate to put work on me 👍

Also happy to see the retweet 😁 
https://twitter.com/threejs_org/status/1369421471643271172

**Review by Mugen87 — DISMISSED** [review]

**Comment by marcofugaro** [comment]
@zalo does the `ShadowmappableMesh.glb` has to be created in any particular way from blender or it can be just a regular model?

**Comment by zalo** [comment]
@supereggbert I _think_ this was due to a WebGL1 compatibility breaking issue I just fixed in my latest commit.   But just to be safe, I added Android to the list of `userAgents` that trigger Half precision floating point textures.  The difference in quality is almost negligible, but there is more support for it on older phones.

@marcofugaro _Technically_ the only requirement is that the object is UV mapped.  However, you'll get oddly colored lines at the texture seams if two faces share an edge in the map (bilinear interpolation edge bleed), so adding some padding is ideal.  I created this model with [automatic UVs from my own program](https://github.com/zalo/CascadeStudio/pull/83), so I'm not sure if Blender has any special needs...

@Mugen87 Thank you for the suggestion to force a WebGL1 Renderer!  I'd been using `texture` instead of `texture2D` 🙃 
Correcting this fixes support on iOS (though, I don't have any Android phones to test with).  I also fixed the warnings you reported.

**Comment by supereggbert** [comment]
> (though, I don't have any Android phones to test with)

Working here now

**Review by Mugen87 — APPROVED** [review]

**Comment by zalo** [comment]
Slightly off-topic: I'm testing out radiosity using [Approximate Radiosity by Stochastic Depth Buffering](http://xevius.org/papers/approxradpaper2010.pdf) (which is technically the multi-bounce extension of this PR's technique (that also doesn't require depth peeling)).

Early results are here:
https://raw.githack.com/zalo/three.js/progressive-radiosity/examples/?q=progress#webgl_shadowmap_progressive

It's not really appropriate for this PR or class.  It's like triple the complexity+compute, converges way more slowly, and it works best with emissive meshes rather than directional lights... all for dubious additional visual quality.  That said, I thought you guys might enjoy it.  :-)

**Comment by mrdoob** [comment]
> That said, I thought you guys might enjoy it. :-)

Yes! 😍

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**